pacific: client: issue a cap release immediately if no cap exists

lxbsz opened the pull request:
backport tracker: https://tracker.ceph.com/issues/61166

---

backport of https://github.com/ceph/ceph/pull/51500 and https://github.com/ceph/ceph/pull/52176
parent tracker: https://tracker.ceph.com/issues/57244 and https://tracker.ceph.com/issues/61781



<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

mchangir commented on src/test/libcephfs/multiclient.cc:
```diff
@@ -97,3 +98,82 @@ TEST(LibCephFS, MulticlientHoleEOF) {
+  while (!(*stop)) {
+    // setattr will make the MDS to acquire xlock for the filelock and
+    // force to revoke caps from clients
+    struct ceph_statx stx = {.stx_size = 0};
```
the above line is causing:
```
/build/ceph-16.2.13-474-gf3910b17/src/test/libcephfs/multiclient.cc:158:43: sorry, unimplemented: non-trivial designated initializers not supported
     struct ceph_statx stx = {.stx_size = 0};
```

for pacific builds.
According to gcc.gnu.org, this feature is available in c++20.
But pacific is built with -std=c++1z


FYI - [build console logs](https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=bionic,DIST=bionic,MACHINE_SIZE=gigantic/71992//consoleFull)

lxbsz replied:
Why I didn't see it in this tracker ?

lxbsz replied:
Fixed it. Please review @mchangir 
Thanks.

vshankar replied:
Seems to be causing failures again: https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=bionic,DIST=bionic,MACHINE_SIZE=gigantic/73151//consoleFull

```
/build/ceph-16.2.14-100-gb009ec3d/src/test/libcephfs/multiclient.cc: In function 'void setattr_func(bool*)':
/build/ceph-16.2.14-100-gb009ec3d/src/test/libcephfs/multiclient.cc:158:43: sorry, unimplemented: non-trivial designated initializers not supported
     struct ceph_statx stx = {.stx_size = 0};
```

@lxbsz

lxbsz replied:
Done. I have fixed it before, not sure why it back again :-(

lxbsz commented:
jenkins test make check

mchangir reviewed (commented):
looks good to me

ljflores commented:
Rados approved: https://tracker.ceph.com/projects/rados/wiki/PACIFIC#httpstrellocomcqRJogcXu1827-wip-yuri2-testing-2023-08-16-1142-pacific

ljflores commented:
Rados approved: https://tracker.ceph.com/projects/rados/wiki/PACIFIC#httpstrellocomc4TiiTR0k1821-wip-yuri7-testing-2023-08-16-1309-pacific-old-wip-yuri7-testing-2023-08-16-0933-pacific-old-wip-yuri7-testing-2023-08-15-0741-pa

batrick commented:
See also in ceph-fuse: /teuthology/yuriw-2023-08-17_14:36:27-fs-wip-yuri7-testing-2023-08-16-1309-pacific-distro-default-smithi/7371813/teuthology.log

```
2023-08-17T15:32:14.007 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr:/build/ceph-16.2.13-531-ge78f2dc9/src/client/Client.cc: In function 'void Client::add_update_cap(Inode*, MetaSession*, uint64_t, unsigned int, unsigned int, unsigned int, unsigned int, inodeno_t, int, const UserPerm&)' thread 7f6bceffd700 time 2023-08-17T15:32:14.005729+0000
2023-08-17T15:32:14.008 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr:/build/ceph-16.2.13-531-ge78f2dc9/src/client/Client.cc: 4393: FAILED ceph_assert(cap.cap_id == cap_id)
2023-08-17T15:32:14.010 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: ceph version 16.2.13-531-ge78f2dc9 (e78f2dc97637da188e6292122efedae3d18948ca) pacific (stable)
2023-08-17T15:32:14.010 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x152) [0x7f6be871ad75]
2023-08-17T15:32:14.010 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 2: /usr/lib/ceph/libceph-common.so.2(+0x267f7d) [0x7f6be871af7d]
2023-08-17T15:32:14.010 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 3: (Client::add_update_cap(Inode*, MetaSession*, unsigned long, unsigned int, unsigned int, unsigned int, unsigned int, inodeno_t, int, UserPerm const&)+0xb9a) [0x55988244b77a]
2023-08-17T15:32:14.010 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 4: (Client::add_update_inode(InodeStat*, utime_t, MetaSession*, UserPerm const&)+0x100b) [0x55988245a6fb]
2023-08-17T15:32:14.010 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 5: (Client::insert_trace(MetaRequest*, MetaSession*)+0xb93) [0x55988245e583]
2023-08-17T15:32:14.011 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 6: (Client::handle_client_reply(boost::intrusive_ptr<MClientReply const> const&)+0x2e4) [0x55988246b024]
2023-08-17T15:32:14.011 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 7: (Client::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0x58d) [0x5598824bbcdd]
2023-08-17T15:32:14.011 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 8: (Messenger::ms_deliver_dispatch(boost::intrusive_ptr<Message> const&)+0x460) [0x7f6be896cd80]
2023-08-17T15:32:14.011 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 9: (DispatchQueue::entry()+0x58f) [0x7f6be896a61f]
2023-08-17T15:32:14.011 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 10: (DispatchQueue::DispatchThread::entry()+0x11) [0x7f6be8a393a1]
2023-08-17T15:32:14.011 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 11: /lib/x86_64-linux-gnu/libpthread.so.0(+0x8609) [0x7f6be8498609]
2023-08-17T15:32:14.011 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 12: clone()
```

lxbsz commented:
> See also in ceph-fuse: /teuthology/yuriw-2023-08-17_14:36:27-fs-wip-yuri7-testing-2023-08-16-1309-pacific-distro-default-smithi/7371813/teuthology.log
> 
> ```
> 2023-08-17T15:32:14.007 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr:/build/ceph-16.2.13-531-ge78f2dc9/src/client/Client.cc: In function 'void Client::add_update_cap(Inode*, MetaSession*, uint64_t, unsigned int, unsigned int, unsigned int, unsigned int, inodeno_t, int, const UserPerm&)' thread 7f6bceffd700 time 2023-08-17T15:32:14.005729+0000
> 2023-08-17T15:32:14.008 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr:/build/ceph-16.2.13-531-ge78f2dc9/src/client/Client.cc: 4393: FAILED ceph_assert(cap.cap_id == cap_id)
> 2023-08-17T15:32:14.010 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: ceph version 16.2.13-531-ge78f2dc9 (e78f2dc97637da188e6292122efedae3d18948ca) pacific (stable)
> 2023-08-17T15:32:14.010 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x152) [0x7f6be871ad75]
> 2023-08-17T15:32:14.010 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 2: /usr/lib/ceph/libceph-common.so.2(+0x267f7d) [0x7f6be871af7d]
> 2023-08-17T15:32:14.010 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 3: (Client::add_update_cap(Inode*, MetaSession*, unsigned long, unsigned int, unsigned int, unsigned int, unsigned int, inodeno_t, int, UserPerm const&)+0xb9a) [0x55988244b77a]
> 2023-08-17T15:32:14.010 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 4: (Client::add_update_inode(InodeStat*, utime_t, MetaSession*, UserPerm const&)+0x100b) [0x55988245a6fb]
> 2023-08-17T15:32:14.010 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 5: (Client::insert_trace(MetaRequest*, MetaSession*)+0xb93) [0x55988245e583]
> 2023-08-17T15:32:14.011 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 6: (Client::handle_client_reply(boost::intrusive_ptr<MClientReply const> const&)+0x2e4) [0x55988246b024]
> 2023-08-17T15:32:14.011 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 7: (Client::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0x58d) [0x5598824bbcdd]
> 2023-08-17T15:32:14.011 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 8: (Messenger::ms_deliver_dispatch(boost::intrusive_ptr<Message> const&)+0x460) [0x7f6be896cd80]
> 2023-08-17T15:32:14.011 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 9: (DispatchQueue::entry()+0x58f) [0x7f6be896a61f]
> 2023-08-17T15:32:14.011 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 10: (DispatchQueue::DispatchThread::entry()+0x11) [0x7f6be8a393a1]
> 2023-08-17T15:32:14.011 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 11: /lib/x86_64-linux-gnu/libpthread.so.0(+0x8609) [0x7f6be8498609]
> 2023-08-17T15:32:14.011 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi008.stderr: 12: clone()
> ```

Let me check this. Thanks @batrick

lxbsz commented:
> Wondering if this is causing: https://tracker.ceph.com/issues/62499

@batrick You are correct and I have taken the tracker and added on comment on it. I will figure out one way to fix it.

Thanks

lxbsz commented:
> > Wondering if this is causing: https://tracker.ceph.com/issues/62499
> 
> @batrick You are correct and I have taken the tracker and added on comment on it. I will figure out one way to fix it.
> 
> Thanks

Just cherry-picked the stale commit https://github.com/lxbsz/ceph/commit/9fbde6da076f2d7c8bfde50c3f9e581e967c0df0, which has been reverted and replaced by https://github.com/lxbsz/ceph/commit/7aaf4ba81b978db63b9cb11a90f881196530e5d5.

batrick commented:
> > > Wondering if this is causing: https://tracker.ceph.com/issues/62499
> > 
> > 
> > @batrick You are correct and I have taken the tracker and added on comment on it. I will figure out one way to fix it.
> > Thanks
> 
> Just cherry-picked the stale commit [lxbsz@9fbde6d](https://github.com/lxbsz/ceph/commit/9fbde6da076f2d7c8bfde50c3f9e581e967c0df0), which has been reverted and replaced by [lxbsz@7aaf4ba](https://github.com/lxbsz/ceph/commit/7aaf4ba81b978db63b9cb11a90f881196530e5d5).

Got it. This looks ready for QA again.

Unfortunately, this was forgotten: https://github.com/ceph/ceph/pull/51661#issuecomment-1557382385

Let's make sure someone does that in the future as we would have avoided a lot of hassle.

gregsfortytwo commented:
@lxbsz why was this buggy in the backport but not main?
(Please put answer in Redmine as well.)

lxbsz commented:
> @lxbsz why was this buggy in the backport but not main? (Please put answer in Redmine as well.)

In the `main` branch, my oldest PR https://github.com/ceph/ceph/pull/47752 was merged, but it introduced a regression when I checked the qa test failures then talked with Venky and we revert it in a following PR https://github.com/ceph/ceph/pull/51661 soon after that. And then I pushed these commits in another new PR https://github.com/ceph/ceph/pull/51500 followed.

This is why we didn't hit it in the following qa tests.

There was only one commit https://github.com/ceph/ceph/pull/47752/commits/9fbde6da076f2d7c8bfde50c3f9e581e967c0df0 was replaced by the last PR https://github.com/ceph/ceph/pull/51500/commits/7aaf4ba81b978db63b9cb11a90f881196530e5d5 and all the other commits didn't change.

But when backporting the commits I just searched the first commit from the `main` branch by using the commit title, which just gave the incorrect commit, and then got the other 4 commit sha1 following it directly.

rishabh-d-dave reviewed (approved):
FS approved. QA run was successful - https://tracker.ceph.com/projects/cephfs/wiki/Pacific#8-Nov-2023.

rishabh-d-dave commented:
> FS approved. QA run was successful - https://tracker.ceph.com/projects/cephfs/wiki/Pacific#8-Nov-2023.

@yuriw @vshankar

lxbsz commented:
@batrick Ping.
There is a `change requested` from you blocks it to get merged.
